aqueduct/gateway/views/decorators.py

```diff
@@ -465,3 +523,117 @@ async def wrapper(request: ASGIRequest, *args, **kwargs):
+            log.error(f"Invalid Host header: {host}")
+            return JsonResponse({"error": "Invalid Host header"}, status=421)
+
+        # Validate Origin header against allowed values
+        # Origin can be absent for same-origin requests, so it's only validated if present
+        allowed_origins = getattr(settings, "MCP_ALLOWED_ORIGINS", [])
+        origin = request.headers.get("origin")
+        if origin:
+            origin_valid = in_wildcard(origin, allowed_origins)
+            if not origin_valid:
+                log.error(f"Invalid Origin header: {origin}")
+                return JsonResponse({"error": "Invalid Origin header"}, status=403)
+
+        return await view_func(request, *args, **kwargs)
+
+    return wrapper
+
+
+def parse_jsonrpc_message(view_func):
+    @wraps(view_func)
+    async def wrapper(request: ASGIRequest, *args, **kwargs):
+        session_id = request.headers.get("Mcp-Session-Id")
+        kwargs["session_id"] = session_id
+
+        if request.method != "POST":
+            if not session_id:
+                log.error(f"Session ID required for MCP server '{kwargs.get('name')}'")
+                return JsonResponse({"error": "Mcp-Session-Id header required"}, status=400)
+
+            return await view_func(request, request_log=None, *args, **kwargs)
+
+        try:
+            body = request.body
+            if body is None:
+                log.error("Request body is None")
+                return JsonResponse({"error": "Missing request body"}, status=400)
+            data = json.loads(body)
+
+            try:
+                json_rpc_message = JSONRPCMessage.model_validate(data)
+            except ValidationError as e:
+                log.error(f"JSON-RPC validation error: {str(e)}, data was: {data!r}")
+                return JsonResponse({"error": f"Invalid JSON-RPC message: {str(e)}"}, status=400)
+            except Exception as e:
+                log.error(f"JSON-RPC parse error: {str(e)}, data was: {data!r}")
+                return JsonResponse(
+                    {"error": f"Failed to parse JSON-RPC message: {str(e)}"}, status=400
+                )
+
+            is_initialize = (
+                hasattr(json_rpc_message.root, "method")
+                and json_rpc_message.root.method == "initialize"
+            )
+
+            if not is_initialize and not session_id:
+                log.error(f"Session ID required for MCP server '{kwargs.get('name')}'")
+                return JsonResponse({"error": "Mcp-Session-Id header required"}, status=400)
+
+            kwargs["json_rpc_message"] = json_rpc_message
+            kwargs["is_initialize"] = is_initialize
+
+            return await view_func(request, *args, **kwargs)
+        except json.JSONDecodeError as e:
+            log.error(f"JSON decode error: {str(e)}, body was: {request.body!r}")
+            return JsonResponse({"error": f"Invalid JSON: {str(e)}"}, status=400)
+        except Exception as e:
```

Comment:
And it doesn't seem to me like this `except Exception` is needed; you check for anything that can go wrong, or catch other exceptions.